### ISSUE TITLE
Fibonacci example and refinement checking

### ISSUE THREAD

**Issue by ruza-net** [linkdd/letlang#2]
Hi, this language is super exciting!

However, I think I'm missing something in the Fibonacci example. Since `natural` is defined by `n > 0`, then `0` is not `natural`. Then in the invocation of `fib(2)`, we need to evaluate `fib(1) - fib(0)`. Since `fib` takes a `natural`, the expression is ill-typed.

How can this example typecheck?

**Comment by linkdd** [comment]
Hello!

Thank you for your comment :) That's a good catch.

This seems to be a bug in the implementation present in the `main` branch, which is a messy Python implementation.

Most work happening at the moment is in the branch `rust-compiler`, which removed some features (I was too ambitious), replaced/redesigned some others as well (to make the language more coherent).

Because of this, the fibonacci example does not compile yet in that branch, but once it does, I'll make sure such case does not happen.

TL;DR:

> How can this example typecheck

It should not

**Comment by ruza-net** [comment]
I see :) Good luck with the new implementation; if you'd like, I could help you somewhat when I have time.

I'm sorry to see the equation solver get dropped; hope that it's not permanent :)

**Comment by ruza-net** [comment]
Great! I'll see when I can contribute something (I'm part-time employed and the semester is starting, so...).
